src/symmetric.rs

```diff
@@ -471,6 +654,147 @@ impl SymmetricAlgorithmKey {
+        ///
+        /// [`BlockCipherKey`]: struct.BlockCipherKey.html
+        /// [`block_cipher::BlockCipher`]: ../../block_cipher/trait.BlockCipher.html
+        pub fn try_into_block_cipher(self) -> Result<BlockCipherKey<A, B>, Self> {
+            let name = self
+                .inner
+                .handle
+                .get_property_unsized::<property::ChainingMode>()
+                .expect("Key to always know its algorithm name");
+            // Unfortunately Windows sometimes insists on returning concatenated
+            // identifiers with \0 in between (so we decode it manually),
+            // for example: ChainingModeECB\u{0}ChainingModeCBC\u{0}".
+            let mode: String = std::char::decode_utf16(name.iter().cloned())
+                .map(|c| c.unwrap_or(std::char::REPLACEMENT_CHARACTER))
+                .collect();
+            if mode.starts_with(ChainingMode::Ecb.to_str()) {
+                Ok(BlockCipherKey {
+                    key: self,
+                    _algo: PhantomData,
+                    _bits: PhantomData,
+                })
+            } else {
+                Err(self)
+            }
+        }
+    }
+
+    use block_cipher::generic_array::{typenum, ArrayLength};
+    use block_cipher::{self, Block, BlockCipher, Key, NewBlockCipher};
+
+    impl<T: typenum::Unsigned> KeyBits for T {
+        const VALUE: Option<usize> = Some(Self::USIZE);
+    }
+
+    impl<B: KeyBits> NewBlockCipher for BlockCipherKey<Aes, B>
+    where
+        B: typenum::Unsigned,
+        // Fancy way of allowing only {128, 192, 256}
+        B: typenum::IsGreaterOrEqual<typenum::U128, Output = typenum::B1>,
+        B: typenum::IsLessOrEqual<typenum::U256, Output = typenum::B1>,
+        B: typenum::PartialDiv<typenum::U64>,
+        // Help the trait solver see that it's also divisible by 8
+        B: typenum::PartialDiv<typenum::U8>,
+        <B as typenum::PartialDiv<typenum::U8>>::Output: ArrayLength<u8>,
+    {
+        /// Key size in bytes with which cipher guaranteed to be initialized.
+        type KeySize = <B as typenum::PartialDiv<typenum::U8>>::Output;
+
+        /// Create new block cipher instance from key with fixed size.
+        fn new(key: &Key<Self>) -> Self {
+            // NOTE: We specifically use ECB chaining mode coupled with empty IV
+            // (in subsequent `{encrypt, decrypt}_block` impls) to provide a
+            // transparent block cipher provided by the CNG API.
+            let prov = SymmetricAlgorithm::open(Aes.id(), ChainingMode::Ecb).unwrap();
+            let key = prov.new_key(key).unwrap();
+            match super::Key::try_from(key) {
+                Ok(key) => Self {
+                    key,
+                    _algo: PhantomData,
+                    _bits: PhantomData,
+                },
+                Err(..) => panic!(),
+            }
+        }
+    }
+
+    impl<B: KeyBits> BlockCipher for BlockCipherKey<Aes, B> {
```

Comment:
Before merging this PR, should we implement `BlockCipher` for other algorithms than AES? My understanding of the code is that we can create a `BlockCipherKey` from any algorithm with `try_into_block_cipher`, but only AES actually implements the `BlockCipher` trait, right?

Reply:
That's right. I implemented `(New)BlockCipher` for remaining block ciphers in https://github.com/emgre/win-crypto-ng/pull/31/commits/2b54a6d6e5261259894e62374e060d1367716570 and https://github.com/emgre/win-crypto-ng/pull/31/commits/364361d74a95dafbb6e29295311e9a455f03d1d8.

Reply:
This way, you can instantiate the `BlockCipher`-compatible key via either
1. `BlockCipherKey::<Algo, Bits>::new(...)` (using `block_cipher::NewBlockCipher`) or
2. `Key::try_into_block_cipher`, `Key` can be obtained via `TryFrom` from a dynamic `SymmetricAlgorithmKey`

Because of the design of `BlockCipher`, you're generally encouraged AIUI to be generic over ciphers implementing *both* `BlockCipher` and `NewBlockCipher`, and so if we wouldn't do that we'd lose quite a bit of compatibility across the crypto ecosystem.

In general, it's good to have a strictly typed (keysize-wise) constructor for block ciphers but the fact that this adds a 3rd public symmetric key type might be slightly confusing, rather than just having a compile-time `Key` and dynamic `SymmetricAlgorithmKey`. The users should ideally be encouraged to use those two and only use `Key::try_into_block_cipher` if they wish to have a `BlockCipher`-capable key.

I think the benefit outweights the possible confusion in this case, and we could return to this later at some point, possibly conflating `BlockCipherKey` with `Key` (by exposing the chaining mode as a type parameter).